### ISSUE TITLE
CSS Order is different when extractCSS is used

### ISSUE THREAD

**Issue by AndrewBogdanovTSS** [nuxt/nuxt.js#4219]
### Version

[2.2.0](https://github.com/nuxt.js/releases/tag/2.2.0)

### Reproduction link

[https://github.com/AndrewBogdanovTSS/nuxt-css-order](https://github.com/AndrewBogdanovTSS/nuxt-css-order)

### Steps to reproduce

- Clone the repo
- $ nuxt build
- $ nuxt start

### What is expected ?

Swiper active control has orange color and positioned at the bottom http://take.ms/UsNSR

### What is actually happening?

Swiper active control has blue color and has a shift of 10px
http://take.ms/ZDjz0

### Additional comments?

This issue is happening due to a fact that order of extracted css files is wrong when extractCSS is in use. 
`vendors.css` is placed after `app.css`  this is causing vendor styles to override stylesheets defined by the user
http://take.ms/ynVEVK 
http://take.ms/D8FH9

<div align="right"><sub><em>This bug report is available on <a href="https://cmty.app/nuxt">Nuxt</a> community (<a href="https://cmty.app/nuxt/nuxt.js/issues/c8052">#c8052</a>)</em></sub></div>

**Comment by bjunc** [comment]
I noticed this as well.  It messes with styling when the selector order is what determines specificity.  For instance, we use tiny reset.  That should be the very first CSS loaded, but in our case (when using extractCSS), it wasn't – making `extractCSS` completely unusable.

**Comment by nmfzone** [comment]
Whopps. I'm not sure, but I think this issue is solved in Nuxt 2.3.1.
I'm using Nuxt 2.2.0 when experiencing this issue, but after I rebuild node_modules and lock file, the issue has gone.

**Comment by nmfzone** [comment]
Still..

**Comment by manniL** [comment]
@nmfzone even with latest nuxt-edge?

**Comment by manniL** [comment]
Looks like it's fixed (https://codesandbox.io/s/github/AndrewBogdanovTSS/nuxt-css-order/tree/master/)

**Comment by nmfzone** [comment]
@manniL Sorry for coming late. I'm not using Nuxt Edge, because it's production environment. So, I must update/upgrade the Vue to make it works?

**Comment by manniL** [comment]
@nmfzone Updating Vue should be fine, yes (`yarn upgrade` or `npm update`)

**Comment by AndrewBogdanovTSS** [comment]
I checked it on my products and I can confirm that the bug is fixed! Thanks @mannil

**Comment by AndrewBogdanovTSS** [comment]
This issue reappeared on version 2.3.4, also was able to reproduce it on Nuxt-edge aka 2.4
Repro link: https://codesandbox.io/s/qqp1wq28yw

**Comment by manniL** [comment]
@AndrewBogdanovTSS Any chance that you could track down the commit that broke it? 🤔

**Comment by AndrewBogdanovTSS** [comment]
@manniL I will look for it, though as far as I remember @pi0 was writing about some updates regarding `mini-css-extract-plugin` in the latest releases. I think it should definitely be related to `mini-css-extract-plugin`

**Comment by manniL** [comment]
@AndrewBogdanovTSS That'd be great!

We switched to https://github.com/webpack-contrib/extract-text-webpack-plugin in the mean time ☺️ 

And the fix came with one of the Vue release IIRC.

**Comment by manniL** [comment]
@AndrewBogdanovTSS maybe https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/586 🤔

**Comment by AndrewBogdanovTSS** [comment]
@manniL yes, it may be related as at least 5 more issues I saw when I was looking for some kind of a solution to this problem :)

**Comment by ruudboon** [comment]
Seeing the same issue with latest releases.

**Comment by thomasaull** [comment]
Anything new to this problem?

**Comment by AndrewBogdanovTSS** [comment]
I just upgraded my project to Nuxt 2.4.2 / Vue 2.6.2 and it seems that problem is fixed again. Although there is still one strange issue: http://take.ms/G6mzJ
as you can see there is 2 identical records for /_nuxt/app.87370d9c34e5a92a0fb0.css file, both for "preload" and regular link. There should be only 2 references, not 4

**Comment by manniL** [comment]
@nmfzone but it's the same CSS 2x in preload and 2x "normal"
@AndrewBogdanovTSS is it the *exact* same link?

**Comment by nmfzone** [comment]
@manniL Whoops, my bad..`revert` comment...

**Comment by AndrewBogdanovTSS** [comment]
@mannil of course it's exactly the same, that's what I wrote in my comment. Right now what I get in the dev tools is application of the same styles twice. It doesn't change visuals but slows down performance for no good reason

**Comment by manniL** [comment]
@AndrewBogdanovTSS Could you open file a dedicated issue for it?

**Comment by AndrewBogdanovTSS** [comment]
@manniL sure

**Comment by manniL** [comment]
Closing here as the initial issue seems to be gone.

**Comment by AndrewBogdanovTSS** [comment]
@manniL  could you reopen this issue? I tried deploying several times and discovered that order is still random. Sometimes it works as expected, but other times it's not. The order I'm currently get is:
1. vendors.css
2. app.css
3. pages.css
4. app.css
https://take.ms/hyZSH
that last one shouldn't be there at all and brakes the whole sequence

**Comment by LostCrew** [comment]
We also most certainly experienced the same issue on Nuxt 2.3.4 and `nuxt-sass-resources-loader` 2.0.5.

**Comment by AndrewBogdanovTSS** [comment]
Any updates regarding this issue?

**Comment by mrleblanc101** [comment]
I have this issue, sometime the scoped style are inlined before the Global CSS from nuxt.config.json

**Comment by poweron707** [comment]
I also found similar issue: 
In my case: I installed Nuxt v2.6.2 and Vuetify v1.5.12
**Case 1:** In nuxt.config.js, Add custom theme(theme.styl) and set "extractCSS: false"  
![case1_setting](https://user-images.githubusercontent.com/1834795/56459117-bd765e80-63c1-11e9-9284-086b4a2f41f7.PNG)

Result is that theme.styl is loaded first, default buttons.styl loaded last, so custom style in theme.styl cannot override default button style ; 
![case1_result](https://user-images.githubusercontent.com/1834795/56459046-17c2ef80-63c1-11e9-9f03-c295d4d900c5.PNG)

**Case 2:** In nuxt.config.js, set "extractCSS: true" 
![case2_setting](https://user-images.githubusercontent.com/1834795/56459260-a2a4e980-63c3-11e9-86cf-a5b2cf16e730.PNG)

![case2_result](https://user-images.githubusercontent.com/1834795/56459272-b8b2aa00-63c3-11e9-877a-1767977b176c.PNG)
As you can see, the CSS load order is changed, the default button style can be override; I got confuse whether it is expected result when using extractCSS in build config

**Comment by thomasaull** [comment]
@stale[bot] @poweron707 described a way to reproduce this issue

**Comment by poweron707** [comment]
1. use nuxt with vuetify example https://github.com/nuxt/nuxt.js/tree/dev/examples/with-vuetify
2. create one styl file like theme.styl and overwrite v-btn style like:
.v-btn
    text-transform: lowercase
3. load this file into nuxt.config.js as above captured screens
4. switch the value of extractCSS

**Comment by clementmas** [comment]
I'm having the same issue. I described the problem here: https://github.com/nuxt/nuxt.js/issues/4204#issuecomment-497910208

**Comment by AndrewBogdanovTSS** [comment]
@manniL please, reopen this issue since it's still not fixed. 
The only workaround for this issue currently is to specify extractCSS as
`extractCSS: process.env.NODE_ENV === 'production'`

**Comment by manniL** [comment]
Still likely related to https://github.com/webpack-contrib/mini-css-extract-plugin/issues/202

maybe @ScriptedAlchemy has a clue? :thinking:

**Comment by trylovetom** [comment]
@manniL same here,

**Comment by ScriptedAlchemy** [comment]
Do you get any order warning errors from webpack? I did see a PR which isn’t exactly related to this issue but does enhance how ordering is Forgive me for not reading the thread fully, is this a dev or production issue.

**Comment by AndrewBogdanovTSS** [comment]
@ScriptedAlchemy yes, it's a dev issue. No - no webpack warnings and it's no more an ordering issue, right now the problem is that in DEV mode we got styles ejected from js into separate css files although it should be happening only in production. Please, link to the PR.

**Comment by ScriptedAlchemy** [comment]
mini-css HMR is designed to generate actual css files though, just like in production. Only difference is it can hmr those stylesheets. 

https://github.com/webpack-contrib/mini-css-extract-plugin/issues/188
https://github.com/webpack-contrib/mini-css-extract-plugin/pull/371

**Comment by AndrewBogdanovTSS** [comment]
@ScriptedAlchemy maybe it can, but that doesn't happening, HMR is not triggering and sourcemaps are broken

**Comment by ScriptedAlchemy** [comment]
Hmmm ive seen heard of sourcemaps issue. Ill pull down your reproduction link and take a look.

**Comment by rnenjoy** [comment]
Im having this issue aswell with the same code ive had before working. Turning off extractCSS gives my css lowest priority vs vuetifys, so mine is overwritten. it has not been like this before."nuxt": "^2.9.2",

**Comment by szafran89** [comment]
I have the same issue, any solution? Thanks in advance!

**Comment by trylovetom** [comment]
my workwround, use node-sass, sass-loader replace extractCSS

Install Dep.
`yarn add node-sass sass-loader --dev`

Your Style
```
<style lang="scss">
...
</style>
```

**Comment by p-moreira** [comment]
Hi, reading this issue, I checked the vuejs docs about merging class attributes. The vuejs docs point me to this behavior:

```
/* MyComponent */
<template>
    <div class="class1">
        <slot />
    </div>
</template>
```

```
<my-component class="class2">
    some content
</my-component>
```

**The result would be:**
```
<div class="class1 class2">
    some content
</div>
```

But, nuxt writes this one:
```
<div class="class2 class1">
    some content
</div>
```

So, first, I need to solve this confusing behavior. Is it a nuxt bug?

`extractCSS: true` - I get wrong class order (different from vuejs docs), but the correct rendered layout
`extractCSS: false` - I get wrong class order (different from vuejs docs) and wrong rendered layout

Someone can tell me what's the correct behavior?

**Comment by AndrewBogdanovTSS** [comment]
@trylovetom I don't understand how scss correlate with extractCSS? I use sass-loader and node-sass and still have same issue

**Comment by ortonomy** [comment]
@trylovetom -- this isn't a solution. 

How did you incorporate sass-loader? did you add a custom webpack plugin?

**Comment by kirosc** [comment]
> 
> 
> @trylovetom I don't understand how scss correlate with extractCSS? I use sass-loader and node-sass and still have same issue

Same here. My mobile version scss code goes before the desktop version but the other way around in the development environment.

**Comment by tegola** [comment]
Same issue here, using Next 2.10.2. I've read all threads related to this issue and I still don't understand what's causing the different extraction order.

I'm trying to keep 1-level selectors, and in the example below the component css is applied after the override in the page (thus is red) because the page css is loaded first.

```html
// component.vue
<template>
    <div class="my-component">
    ...
    </div>
</template>

<style>
.my-component {
    width: 100px;
    height: 100px;
    background-color: red;
}
</style>
```

```html
// page.vue
<template>
    <div class="my-page">
        <my-component class="my-page__component" />
    </div>
</template>

<style>
.my-page__component {
    background-color: blue;
}
</style>
```

**Comment by sanasus** [comment]
> Та же проблема здесь, используя Next 2.10.2. Я прочитал все темы, связанные с этой проблемой, и я до сих пор не понимаю, что вызывает другой порядок извлечения.
> 
> Я пытаюсь сохранить одноуровневые селекторы, и в приведенном ниже примере компонент css применяется после переопределения на странице (таким образом, красного цвета), потому что сначала загружается css страницы.
> 
> ```
> // component.vue
>  < template > 
>     < div  class = " my-component " >
>     ...
>     </ div > 
> </ template >
> 
> < стиль > 
> . my-component {
>      ширина : 100 пикселей ;
>     высота : 100 пикселей ;
>     цвет фона : красный;
> }
> </ style >
> ```
> 
> ```
> // page.vue
>  < template > 
>     < div  class = " my-page " > 
>         < my-component  class = " my-page__component " />
>      </ div > 
> </ template >
> 
> < стиль > 
> . my-page__component {
>      background-color : blue;
> }
> </ style >
> ```

I have exactly the same problem. The import order and so on do not affect this, if I extract everything in 1 css file, then there is no such error, but the console is full of warnings.

**Comment by mateuszs** [comment]
up +1. Any solutions?

@sanasus  @tegola

**Comment by fvonellerts** [comment]
I had the same issue when importing my global SCSS in `layouts/default.vue` instead of adding it to  the `css` Array in `nuxt.config.js`.

**Comment by tegola** [comment]
> up +1. Any solutions?

I haven't found any except using page-scoped styles or two-level selectors.

**Comment by mjmnagy** [comment]
Same issue here - just leaving in dev mode for the time being

**Comment by stouch** [comment]
Same here ....

**Comment by p-moreira** [comment]
Same too.

- import the global CSS in nuxtconfig.js

extractCSS = false:
    - DEV mode:  all is ok
    - PRODUCTION mode: wrong styling is applied

extractCSS = true:
    - all is ok in DEV and in PRODUCTION

**Comment by mayurwish** [comment]
Same here
Development mode: Css & Scss loads well
Production mode: fail to load some css & scss even though package.json includes node-sass, sass-loader & @nuxtjs/style-resources

**Comment by clarkdo** [comment]
Can anyone who has the issue please provide minimum reproducible test repo?

We'll try to figure out what's root cause in extract-css-plugin.

**Comment by tegola** [comment]
> Can anyone who has the issue please provide minimum reproducible test repo?

Here you go:
https://github.com/tegola/nuxt-css-order-test-case

**Comment by aaronransley** [comment]
Checking in on this, just ran into this during a big production deploy 😬

**Comment by kyriediculous** [comment]
I'm currently running into this whereby my scoped styles that are imported through component specific stylesheets get overwritten in production mode. 

Tried this with Nuxt 2.10.2 and 2.13.2 

```
<style="scss" scoped>
 import ...
</style>
```

is there any solution to this? 

setting `extractCss: true` in my `nuxt.config.js` file didn't work for me.

**Comment by Everkers** [comment]
I have the same issue in production : (

**Comment by kyriediculous** [comment]
Create a new class that doesn't exist in any other css

```
.lplogo {
  opacity:75% !important;;
}
```

Apply it 


result is 

```
.lplogo[data-v-b0e3836e] {
    opacity: 1%!important;
}
```


Seriously; what kind of joke is this ?

**Comment by clarkdo** [comment]
@tegola I've looked into your repo, it looks the issue is related to `vue-loader`:

1. In dev mode, `vue-loader` will ignore styles returned from ssr and re-added them as components loading order (`pages/index -> MyComponent`) in browser side, so you can see correct `background-color`
2. In prod mode, `vue-loader` will attach `style injecting` to `beforeCreate` hook of vue component. In Vue, `beforeCreate` and `created` of the parent run first, so `beforeCreate` of `pages/index` will be executed before `MyComponent` and style of `pages/index` is injected before `MyComponent`

I found that there already has some discussions in vue-loader: https://github.com/vuejs/vue-loader/issues/808#issuecomment-303785334 https://github.com/vuejs/vue-loader/issues/521 , I'll talk to team and try to figure a proper solution for it.

**Comment by davidlung** [comment]
The issue with extractCSS -> wrong CSS order still exists and has not been fixed according to the last merge.
NUXT - v2.15.8

**Comment by dumptyd** [comment]
The issue still persists.

In my very specific case, this started happening when I used an auto-imported button component in my default layout. From that point onwards, all of my button styles were being imported **before** the normalize styles and this broke all the buttons in my app.

The solution was to add the button component (among others) to the `components[].ignore` option in my nuxt config and manually importing and registering them as global components. This fixes the issue because the ordering is defined predictably by you.

<details>
  <summary>Relevant snippets</summary>

`nuxt.config.js`
```js
 plugins: [
    '~/components/my-global-components/init.js',
  ],
  components: [
    {
      path: '~/components',
      ignore: ['./my-global-components/*/**.{vue,js}']
    }
  ],
```

`/components/my-global-components/init.js`

```js
import Vue from 'vue';
import Button from './button';

export default (context, inject, ...args) => {
  Vue.component('my-button', Button);
};

```

</details>

**Comment by AndrewBogdanovTSS** [comment]
@dumptyd disabling nuxt/components is not a solution, maybe you could log a bug at https://github.com/nuxt/components/issues ?
